### ISSUE TITLE
inter class center-to-center loss

### ISSUE THREAD

**Issue by ydhongHIT** [edwardyehuang/CAR#4]
Why do you use the softmax and stop the gradient on keys?

**Comment by edwardyehuang** [comment]
1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency. <br>
Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.

2. It is important to fixed the one class when you want to reduce its dependency with other classes.

**Comment by ydhongHIT** [comment]
> 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency. 
>    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> 2. It is important to fixed the one class when you want to reduce its dependency with other classes.

Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?

**Comment by edwardyehuang** [comment]
> > 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency.
> >    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> > 2. It is important to fixed the one class when you want to reduce its dependency with other classes.
> 
> Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?

It is ok to stop grads on the query.

**Comment by ydhongHIT** [comment]
> > > 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency.
> > >    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> > > 2. It is important to fixed the one class when you want to reduce its dependency with other classes.
> > 
> > 
> > Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?
> 
> It is ok to stop grads on the query.

Thanks for your reply. In my opinion, the class center may be zeros. Doesn't it conflict with the softmax? Normally, similarity with zero vector is meaningless but 0 element is meaningful for softmax.

**Comment by edwardyehuang** [comment]
> > > > 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency.
> > > >    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> > > > 2. It is important to fixed the one class when you want to reduce its dependency with other classes.
> > > 
> > > 
> > > Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?
> > 
> > 
> > It is ok to stop grads on the query.
> 
> Thanks for your reply. In my opinion, the class center may be zeros. Doesn't it conflict with the softmax? Normally, similarity with zero vector is meaningless but 0 element is meaningful for softmax.

No, unless all the class centers in the current batch are 0. Note that, our target is to minimize the similarity (inter-class dependency) below thresholds. So, it is okay if some class centers in the current batch are 0 (below the threshold, no need to regularize)

**Comment by ydhongHIT** [comment]
> > > > > 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency.
> > > > >    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> > > > > 2. It is important to fixed the one class when you want to reduce its dependency with other classes.
> > > > 
> > > > 
> > > > Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?
> > > 
> > > 
> > > It is ok to stop grads on the query.
> > 
> > 
> > Thanks for your reply. In my opinion, the class center may be zeros. Doesn't it conflict with the softmax? Normally, similarity with zero vector is meaningless but 0 element is meaningful for softmax.
> 
> No, unless all the class centers in the current batch are 0. Note that, our target is to minimize the similarity (inter-class dependency) below thresholds. So, it is okay if some class centers in the current batch are 0 (below the threshold, no need to regularize)

I don't know if I'm getting this right. If one class center is zero, the inner product between this center and any other center is 0. The softmax value should be 1/self.num_classes, larger than threshold (inter_c2c_loss_threshold/(self.num_classes-1)).

**Comment by edwardyehuang** [comment]
> > > > > > 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency.
> > > > > >    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> > > > > > 2. It is important to fixed the one class when you want to reduce its dependency with other classes.
> > > > > 
> > > > > 
> > > > > Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?
> > > > 
> > > > 
> > > > It is ok to stop grads on the query.
> > > 
> > > 
> > > Thanks for your reply. In my opinion, the class center may be zeros. Doesn't it conflict with the softmax? Normally, similarity with zero vector is meaningless but 0 element is meaningful for softmax.
> > 
> > 
> > No, unless all the class centers in the current batch are 0. Note that, our target is to minimize the similarity (inter-class dependency) below thresholds. So, it is okay if some class centers in the current batch are 0 (below the threshold, no need to regularize)
> 
> I don't know if I'm getting this right. If one class center is zero, the inner product between this center and any other center is 0. The softmax value should be 1/self.num_classes, larger than threshold (inter_c2c_loss_threshold/(self.num_classes-1)).

Note that, we have a scale after softmax https://github.com/edwardyehuang/CAR/blob/master/car_core/utils.py#L70

**Comment by ydhongHIT** [comment]
> > > > > > > 1. We use softmax only because it is the simplest and most commonly used way in self-attention for similarity calculation. Yes, there must be a better way to conduct similarity calculation in our case. However, it is beyond our paper, where we only want to present the importance of regularizing (reducing) inter-class dependency.
> > > > > > >    Note that, it may be hard to find a similarity function compatibles with so many baselines like we present in this paper.
> > > > > > > 2. It is important to fixed the one class when you want to reduce its dependency with other classes.
> > > > > > 
> > > > > > 
> > > > > > Thanks for your reply. Does the softmax break the symmetry? For example, is it wrong to stop the gradient on querys?
> > > > > 
> > > > > 
> > > > > It is ok to stop grads on the query.
> > > > 
> > > > 
> > > > Thanks for your reply. In my opinion, the class center may be zeros. Doesn't it conflict with the softmax? Normally, similarity with zero vector is meaningless but 0 element is meaningful for softmax.
> > > 
> > > 
> > > No, unless all the class centers in the current batch are 0. Note that, our target is to minimize the similarity (inter-class dependency) below thresholds. So, it is okay if some class centers in the current batch are 0 (below the threshold, no need to regularize)
> > 
> > 
> > I don't know if I'm getting this right. If one class center is zero, the inner product between this center and any other center is 0. The softmax value should be 1/self.num_classes, larger than threshold (inter_c2c_loss_threshold/(self.num_classes-1)).
> 
> Note that, we have a scale after softmax https://github.com/edwardyehuang/CAR/blob/master/car_core/utils.py#L70

Thanks for your patient reply. But this scale is before softmax, as shown in your code. I hope I read that right.

**Comment by edwardyehuang** [comment]
Oh, I just noticed that. Thank you for pointing out this issue. However, I think there is no effect because class_center = 0 means there is no related feature that will be optimized during the regularization.

**Comment by edwardyehuang** [comment]
https://github.com/edwardyehuang/CAR/blob/master/car_core/car.py#L180

**Comment by ydhongHIT** [comment]
But it indeed affects the gradient. Maybe, it is 

> Oh, I just noticed that. Thank you for pointing out this issue. However, I think there is no effect because class_center = 0 means there is no related feature that will be optimized during the regularization.

But it seems affect the gradient. Anyhow, the experimental results are good. Thanks for your great work.

**Comment by ydhongHIT** [comment]
But it seems affect the gradient. Anyhow, the experimental results are good. Thanks for your great work.

**Comment by edwardyehuang** [comment]
> But it seems affect the gradient. Anyhow, the experimental results are good. Thanks for your great work.

divide_no_nan should not bring the grads to the class with zero pixels.

Thanks for your interest btw.

**Comment by edwardyehuang** [comment]
Ok, I just noticed. This issue may bring some slightly "scale" effects to the class has pixels in the current batch. 

Fortunately, because divide_no_nan brings no grads to the class with zero pixels, this issue should only affect the fitting speed on the training set.

I will explore it in my further research.

**Comment by edwardyehuang** [comment]
> But it seems affect the gradient. Anyhow, the experimental results are good. Thanks for your great work.

Would you like to add my wechat "edwardyehuang" btw?
